vars/dockerExecute.groovy

```diff
@@ -1,53 +1,85 @@
 
-    def STEP_NAME = 'dockerExecute'
-    def PLUGIN_ID_DOCKER_WORKFLOW = 'docker-workflow'
+@Field Set GENERAL_CONFIG_KEYS = ['jenkinsKubernetes']
 
-    handlePipelineStepErrors(stepName: STEP_NAME, stepParameters: parameters){
-        def dockerImage = parameters.dockerImage ?: ''
-        Map dockerEnvVars = parameters.dockerEnvVars ?: [:]
-        def dockerOptions = parameters.dockerOptions ?: ''
-        Map dockerVolumeBind = parameters.dockerVolumeBind ?: [:]
+@Field Set PARAMETER_KEYS = ['dockerImage',
+                             'dockerOptions',
+                             'dockerWorkspace',
+                             'dockerEnvVars',
+                             'dockerVolumeBind']
+@Field Set STEP_CONFIG_KEYS = PARAMETER_KEYS
 
-        if(dockerImage) {
+void call(Map parameters = [:], body) {
+    handlePipelineStepErrors(stepName: STEP_NAME, stepParameters: parameters) {
+        final script = parameters.script
+        if (script == null)
+            script = [commonPipelineEnvironment: commonPipelineEnvironment]
 
-            if (! isPluginActive(PLUGIN_ID_DOCKER_WORKFLOW) ) {
-                echo "[WARNING][${STEP_NAME}] Docker not supported. Plugin '${PLUGIN_ID_DOCKER_WORKFLOW}' is not installed or not active. Configured docker image '${dockerImage}' will not be used."
-                dockerImage = null
+        Map config = ConfigurationHelper
+            .loadStepDefaults(this)
+            .mixinGeneralConfig(script.commonPipelineEnvironment, GENERAL_CONFIG_KEYS)
+            .mixinStageConfig(script.commonPipelineEnvironment, parameters.stageName?:env.STAGE_NAME, STEP_CONFIG_KEYS)
+            .mixinStepConfig(script.commonPipelineEnvironment, STEP_CONFIG_KEYS)
+            .mixin(parameters, PARAMETER_KEYS)
+            .withMandatoryProperty('dockerImage')
+            .use()
+
+        if (isKubernetes() && config.dockerImage) {
+            if (env.POD_NAME && isContainerDefined(config)) {
+                container(getContainerDefined(config)) {
+                    echo "Executing inside a Kubernetes Container"
+                    body()
+                    sh "chown -R 1000:1000 ."
+                }
+            } else {
+                dockerExecuteOnKubernetes(
+                    script: script,
+                    dockerImage: config.dockerImage,
+                    dockerEnvVars: config.dockerEnvVars,
+                    dockerWorkspace: config.dockerWorkspace
+                ){
+                    echo "Executing inside a Kubernetes Pod"
+                    body()
+                }
+            }
+        } else {
+            if (!JenkinsUtils.isPluginActive(PLUGIN_ID_DOCKER_WORKFLOW)) {
+                echo "[WARNING][${STEP_NAME}] Docker not supported. Plugin '${PLUGIN_ID_DOCKER_WORKFLOW}' is not installed or not active. Configured docker image '${config.dockerImage}' will not be used."
+                config.dockerImage = null
```

Comment:
I personally doubt it is a good idea to modify the config outside the ConfigurationHelper etc. In case we need to null something it is IMO better to assign the value to a variable and to null that variable rather than working on the configuration itself (applies also elsewhere  in that code)

Reply:
Maybe just a boolean useDocker=false